File: algorithms/cpp/402.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string removeKdigits(string num, int k) {
        int digits = num.size()-k;
        vector<char> s(num.size());
        int top = 0;
        for (int i = 0; i < num.size(); i++) {
            while (top > 0 && s[top-1] > num[i] && k > 0) {
                top -= 1;
                k -= 1;
            }
            s[top++] = num[i];
        }
        int idx = 0;
        while (idx < digits && s[idx] == '0') idx += 1;
        if (idx == digits) return "0";
        return string(s.begin()+idx, s.begin()+digits);
    }
};

int main() {
    Solution solution;
    cout << solution.removeKdigits("10", 1) << endl;
    return 0;
}

File: algorithms/python3/402.py
class Solution:
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        rls = self.helper(num, k).lstrip('0')
        return rls if len(rls) else '0'

    def helper(self, num, k):
        if k == 0:
            return num
        for i in range(len(num)-1):
            if num[i] > num[i+1]:
                return self.helper(num[:i] + num[i+1:], k-1)
        return self.helper(num[:-1], k-1)
